### PR TITLE
BREAKING_CHANGE: change format for binary classification heads

### LINKED ISSUE #374: predict_type="prob" does not work with out_features=1
Hi @sebffischer I expected that I should be able to use measures to evaluate predictions, for binary classification with a torch learner with out_features=1. Here is an example adapted from #373 (thanks!)
```r
library(mlr3torch)
nn_bce_loss3 = nn_module(c("nn_bce_with_logits_loss3", "nn_loss"),
  initialize = function(weight = NULL, reduction = "mean", pos_weight = NULL) {
    self$loss = nn_bce_with_logits_loss(weight, reduction, pos_weight)
  },
  forward = function(input, target) {
    self$loss(input$reshape(-1), target$to(dtype = torch_float())-1)
  }
)
loss = nn_bce_loss3()
loss(torch_randn(10, 1), torch_randint(0, 1, 10))
task = tsk("sonar")
graph = po("torch_ingress_num") %>>%
  nn("linear", out_features = 1) %>>%
  po("torch_loss", loss = nn_bce_loss3) %>>%
  po("torch_optimizer") %>>%
  po("torch_model_classif",
     epochs = 1,
     batch_size = 32,
     predict_type="prob")
glrn = as_learner(graph)
glrn$train(task)
glrn$predict(task)
```
The code above has predict_type="prob" and out_features=1 so I am getting the following error on current main
```r
if(FALSE){#broke
  remotes::install_github("mlr-org/mlr3torch@6e99e02908788275622a7b723d211f357081699a")
  glrn$predict(task)
  ## Erreur dans dimnames(x) <- dn : 
  ##   la longueur de 'dimnames' [2] n'est pas égale à l'étendue du tableau
  ## This happened PipeOp torch_model_classif's $predict()
}
```
The error happens because the torch model outputs only one column, but some later code assumes there are two.

I hacked a solution that fixes this (see below), and I will file a PR.
```r
if(FALSE){#fix
  remotes::install_github("tdhock/mlr3torch@69d4adda7a71c05403d561bf3bb1ffb279978d0d")
  glrn$predict(task)
  ## <PredictionClassif> for 208 observations:
  ##  row_ids truth response
  ##        1     R        M
  ##        2     R        M
  ##        3     R        M
  ##      ---   ---      ---
  ##      206     M        M
  ##      207     M        M
  ##      208     M        M
}
```

### PR DESCRIPTION
**Pull request by sebffischer**
Resolves #374 

@tdhock With this PR, `mlr3torch` will now expect an output of shape `(batch_size, 1)` for binary classification problems.
`t_loss("cross_entropy")` will automatically select the appropriate loss depending on the number of classes.

In the example below, you see it in action.
We train the same learner once on a binary classification problem (iris subset to two classes) and on a multi-class classification problem (iris).

Depending on the task, the correct output dimension will be set, the targets will be loaded correctly during training, and the correct loss function is instantiated. This is now also respected by all `mlr3torch` learners.

``` r
library(mlr3torch)

cb = torch_callback("loss_fn",
  state_dict = function() {
    self$ctx$loss_fn
  },
  load_state_dict = function(state_dict) {
    self$x = x
  }
)

learner = lrn("classif.mlp", neurons = 100, batch_size = 32, epochs = 10,
  callbacks = cb)

tsk_binary = tsk("iris")$filter(1:100)$droplevels()
tsk_multi = tsk("iris")

learner$train(tsk_binary)
learner$network(torch_randn(1, 4))
#> torch_tensor
#> -0.3322
#> [ CPUFloatType{1,1} ][ grad_fn = <AddmmBackward0> ]
class(learner$model$callbacks$loss_fn)
#> [1] "nn_bce_with_logits_loss" "nn_loss"                
#> [3] "nn_module"

learner$train(tsk_multi)
learner$network(torch_randn(1, 4))
#> torch_tensor
#>  1.0422  0.3267 -0.5658
#> [ CPUFloatType{1,3} ][ grad_fn = <AddmmBackward0> ]
class(learner$model$callbacks$loss_fn)
#> [1] "nn_cross_entropy_loss" "nn_weighted_loss"      "nn_loss"              
#> [4] "nn_module"
```

<sup>Created on 2025-04-16 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Note that after this is is **not** possible to have a neural network with output dim `(batch_size, 2)` for binary classification. I think this is okay, as `(batch_size, 1)` is clearly better.

``` r
glrn = po("torch_ingress_num") %>>%
  nn("linear", out_features = 2) %>>%
  po("torch_loss", loss = "cross_entropy") %>>%
  po("torch_optimizer", "adamw") %>>%
  po("torch_model_classif", epochs = 1, batch_size = 32) |> as_learner()

glrn$train(tsk("sonar"))
#> Error in (function (self, target, weight, pos_weight, reduction) : output with shape [32, 1] doesn't match the broadcast shape [32, 2]
#> Exception raised from mark_resize_outputs at /Users/runner/work/libtorch-mac-m1/libtorch-mac-m1/pytorch/aten/src/ATen/TensorIterator.cpp:1208 (most recent call first):
#> frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 52 (0x10422811c in libc10.dylib)
#> frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) + 140 (0x104224d6c in libc10.dylib)
#> ... 
#> This happened PipeOp torch_model_classif's $train()
```

<sup>Created on 2025-04-16 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/LearnerTorch.R [hunk 1]
```diff
@@ -83,6 +83,11 @@
 #'   is trained by the learner.
 #'   For classification, the output of this network are expected to be the scores before the application of the
 #'   final softmax layer.
+#'   The expected output dimenions are as follows:
+#'   * binary classification: `c(batch_size, 1)`
+#'   * multiclass classification: `c(batch_size, n_classes)`
+#'   * regression: `c(batch_size, 1)`
+#'   You can use the [`output_dim_for`] function to obtain the correct output dimension for the task.
```
Comment (sebffischer) [inline R/LearnerTorch.R:90]:
```suggestion
#'   You can use [`output_dim_for()`] to obtain the correct output dimension for a given task.
```

**Comment by tdhock** [comment]
The code seems to work as I would expect, thanks!
I tried this code:
```r
remotes::install_github("mlr-org/mlr3torch@c03d61a18e9785e2dbb5b20e2b6dada74a9b58b8")
stask <- mlr3::tsk("sonar")
po_list <- list(
  mlr3torch::PipeOpTorchIngressNumeric$new(),
  mlr3torch::nn("head"),
  mlr3pipelines::po(
    "torch_loss",
    loss = torch::nn_bce_with_logits_loss),
  mlr3pipelines::po("torch_optimizer"),
  mlr3pipelines::po(
    "torch_model_classif",
    epochs = 1,
    batch_size = 3,
    predict_type="prob"))
graph <- Reduce(mlr3pipelines::concat_graphs, po_list)
glrn <- mlr3::as_learner(graph)
glrn$train(stask)
glrn$predict(stask)
```
And I got this result:
```r
> glrn$predict(stask)
<PredictionClassif> for 208 observations:
 row_ids truth response    prob.M    prob.R
       1     R        M 0.5324456 0.4675544
       2     R        M 0.5387607 0.4612393
       3     R        M 0.5339928 0.4660072
     ---   ---      ---       ---       ---
     206     M        M 0.5342165 0.4657835
     207     M        M 0.5279804 0.4720196
     208     M        M 0.5211176 0.4788824
```

File: man/nn_graph.Rd [hunk 2]
```diff
@@ -23,8 +23,28 @@ Whether output should be a list of tensors. If \code{FALSE} (default), then \cod
 \code{\link{nn_graph}}
 }
 \description{
-Represents a neural network using a \code{\link[mlr3pipelines:Graph]{Graph}} that usually costains mostly \code{\link{PipeOpModule}}s.
+Represents a neural network using a \code{\link[mlr3pipelines:Graph]{Graph}} that costains mostly \code{\link{PipeOpModule}}s.
```
Comment (tdhock) [inline man/nn_graph.Rd:26]: typo -> contains

File: man/output_dim_for.Rd [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+The task.}
+
+\item{...}{(any)\cr
+Additional arguments. Not used yet.}
+}
+\description{
+Calculates the output dimension of a neural network for a given task that is expected by
+\pkg{mlr3torch}.
+For classification, this is the number of classes (unless it is a binary classification task,
+where it is 1). For regression, it is 1.
```
Comment (tdhock) [inline man/output_dim_for.Rd:20]: for regression, this contradicts the docs above https://github.com/mlr-org/mlr3torch/pull/385/files#diff-599122382a84c6d4650caa7d1fdfad391c4825afb169b61e1a24da098c3f3e80
`\item regression, the output dimension is the same as the number of target features.`

File: tests/testthat/helper_autotest.R [hunk 4]
```diff
@@ -458,8 +460,28 @@ expect_learner_torch = function(learner, task, check_man = TRUE, check_id = TRUE
   testthat::expect_true(batch$y$device == torch_device("cpu"))
   if (task$task_type == "regr") {
     testthat::expect_true(batch$y$dtype == torch_float())
-  } else {
+  } else if (task$task_type == "classif" && "twoclass" %in% task$properties) {
+    testthat::expect_true(batch$y$dtype == torch_float())
```
[the remark this answers is not in the document]
Comment (sebffischer) [reply tests/testthat/helper_autotest.R:464]: good catch! I added a section *Network Head and Target Encoding* to the documentation of `LearnerTorch`.

File: R/LearnerTorch.R [hunk 5]
```diff
@@ -29,6 +29,21 @@
...
+#' * binary classification: `c(batch_size, 1)`, representing the logits for the positive class.
+#' * multiclass classification: `c(batch_size, n_classes)`, representing the logits for all classes.
+#' * regression: `c(batch_size, 1)` representing the response prediction.
+#'
+#' Furthermore, the target encoding is expected to be as follows:
+#' * regression: The `numeric` target variable of a [`TaskRegr`][mlr3::TaskRegr] is encoded as a
+#'   [`torch_float`][torch::torch_float] with shape `c(batch_size, 1)`.
+#' * binary classification: The `factor` target variable of a [`TaskClassif`][mlr3::TaskClassif] is encoded as a
+#'   [`torch_float`][torch::torch_float] with shape `(batch_size, 1)` where the positive class is `1` and the negative
+#'   class is `0`.
```
Comment (tdhock) [inline R/LearnerTorch.R:43]: ok, but what is the positive and negative class at the R-level?
In R it is a factor with two levels. Is the first factor level always considered the positive class?

Comment (sebffischer) [reply R/LearnerTorch.R:43]: `mlr3` binary classification tasks have a field `$positive`. Will add that to the docs!

File: R/LearnerTorch.R [hunk 6]
```diff
@@ -29,6 +29,21 @@
...
 #'
+#' @section Network Head and Target Encoding:
+#' Torch learners are expected to have the following output:
+#' * binary classification: `c(batch_size, 1)`, representing the logits for the positive class.
+#' * multiclass classification: `c(batch_size, n_classes)`, representing the logits for all classes.
+#' * regression: `c(batch_size, 1)` representing the response prediction.
+#'
+#' Furthermore, the target encoding is expected to be as follows:
+#' * regression: The `numeric` target variable of a [`TaskRegr`][mlr3::TaskRegr] is encoded as a
+#'   [`torch_float`][torch::torch_float] with shape `c(batch_size, 1)`.
```
[the remark this answers is not in the document]
Comment (sebffischer) [reply R/LearnerTorch.R:40]: Not yet, but we could implement a `Task` where the target is a `lazy_tensor`, which would allow this.

File: R/LearnerTorch.R (hunk 5 above)
Comment (sebffischer) [reply R/LearnerTorch.R:43]: And yes, it is also ensured that the positive class is the first level.

File: R/LearnerTorch.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdhock) [reply R/LearnerTorch.R:40]: setting target to lazy_tensor would be specific to torch, right? other non-torch learners can't use lazy_tensors, right?
it would be better if there would be support for any learners (for example glmnet or rpart) and not just torch.

File: R/LearnerTorch.R (hunk 5 above)
Comment (tdhock) [reply R/LearnerTorch.R:43]: thanks

File: R/LearnerTorch.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (sebffischer) [reply R/LearnerTorch.R:40]: yeah that’s true, if you want this this should be a feature request in mlr3.
